File: codewars/js/4kyu/matrix_determinant.js
function determinant(matrix) {
    if (matrix.length == 1) {
        return matrix[0][0];
    }

    if (matrix.length == 2) {
        return matrix[0][0]*matrix[1][1] - matrix[0][1]*matrix[1][0];
    }

    var result = 0;

    for (var j = 0; j < matrix.length; ++j) {
        result += Math.pow(-1, j) * matrix[0][j] * determinant(submatrix(matrix, j));
    }

    return result;
}

function submatrix(matrix, index) {
    var submatrix = [],
        n = matrix.length;

    for (var i = 1; i < n; ++i) {
        var row = [];

        for (var j = 0; j < n; ++j) {
            if (j != index) {
                row.push(matrix[i][j]);
            }
        }
        submatrix.push(row);
    }

    return submatrix;
}

File: codewars/python/4kyu/matrix_determinant.py
def determinant(matrix):
    assert len(matrix) >= 1

    if len(matrix) == 1:
        return matrix[0][0]

    pivots = enumerate(matrix[0])
    return sum(v * (-1)**i * determinant(submatrix(matrix, i)) for i,v in pivots)


def submatrix(matrix, index):
    return [[v for i,v in enumerate(row) if i != index] for row in matrix[1:]]
